File: FSAdvent/FSAdvent/Y2015/Day1.fs
module FSAdvent.Y2015.Day1
open FSAdvent.Core.Functions

let operationScore operation =
    match operation with
    | '(' -> 1
    | ')' -> -1
    | _ -> 0

let step floor operation =
    floor + operationScore(operation)

let part1 (n:string) =
    n.ToCharArray() |> Seq.fold step 0

let stepIndexed (index, floor) operation =
    (inc index, step floor operation)

let part2 (n:string) =
    n.ToCharArray()
    |> Seq.scan stepIndexed (0, 0)
    |> Seq.find (snd >> isNegative)
    |> fst

File: java/src/main/java/com/brandoncorrea/y2015/Day1.java
package com.brandoncorrea.y2015;

public class Day1 {

    public static int part1(String input) {
        return sanitize(input).chars().map(Day1::step).sum();
    }

    public static int part2(String input) {
        int floor = 0;
        int position = 1;
        for (char c : sanitize(input).toCharArray()) {
            floor += step(c);
            if (floor == -1) return position;
            position++;
        }
        return -1;
    }

    static String sanitize(String input) {
        return input == null ? "" : input;
    }

    static int step(int direction) {
        if (direction == '(') return 1;
        else if (direction == ')') return -1;
        return 0;
    }
}
